test/test_train_mp_mnist.py

```diff
@@ -129,14 +129,16 @@ def train_mnist(flags, **kwargs):
 
   def train_loop_fn(loader):
     tracker = xm.RateTracker()
-
     model.train()
     for step, (data, target) in enumerate(loader):
-      optimizer.zero_grad()
-      output = model(data)
-      loss = loss_fn(output, target)
-      loss.backward()
-      xm.optimizer_step(optimizer)
+      with xp.StepTrace('train_mnist', step_num=step):
+        with xp.Trace('build_graph'):
+          optimizer.zero_grad()
+          output = model(data)
+          loss = loss_fn(output, target)
+          loss.backward()
+        # For profiling purposes we're moving the mark step out of iterator
+        xm.optimizer_step(optimizer, barrier=True)
```

Comment:
why barrier here? is it because we don't mark step and the profiler will spit out wrong info otherwise?

Reply:
I see you removed mark_step from iterator. ok.

This usage feels counter-intuitive to me. Can we handle this in the __enter__, __exit__ methods of the decorator?

I'm not sure if we can make the iterator not mark_step (even if it does, fine, whatever), but we can maybe mark_step at __exit__ ? Otherwise, the more complex steps get, the more bookkeeping the user has to do to mark_step at the right time etc. For example, in dlrm, we have 2 optimizer_step's per model step. WDYT?

Reply:
As discussed offline, I'll see if there are cleaner ways from a UX perspective on the profiling experience for these context manager trace marker APIs in terms of moving around `mark_step` calls.